File: N - Queens/n_queens.cpp
#include<bits/stdc++.h>
using namespace std;

bool isSafe(vector<int> &board, int row, int col) {
    for(int i=0;i<row;i++) {
        if(board[i] == col || abs(board[i] - col) == abs(i - row)) return false;
    }

    return true;
}

void solveNQueens(vector<vector<int>> &solutions, vector<int> &board, int row, int n) {
    if(row == n) {
        solutions.push_back(board);
        return;
    }

    for(int col=0;col<n;col++) {
        if(isSafe(board, row, col) == true) {
            board[row] = col;
            solveNQueens(solutions, board, row + 1, n);
            board[row] = -1;
        }
    }
}

void printBoard(vector<int> solution) {
    int n = solution.size();

    for(int i=0;i<n;i++) {
        for(int j=0;j<n;j++) {
            if(solution[i] == j) {
                cout<<"Q ";
            } else {
                cout<<"* ";
            }
        }
        cout<<endl;
    }
    cout<<endl;
}

void nQueensBackTracking(int n) {
    vector<vector<int>> solutions;
    vector<int> board(n, -1);
    solveNQueens(solutions, board, 0, n);

    cout<<"Number of solutions for "<<n<<"-queens: "<<solutions.size()<<endl;
    for(auto solution : solutions) {
        printBoard(solution);
    }
}

int main() {
    nQueensBackTracking(5);

    return 0;
}

File: N - Queens/n_queens.py
def is_safe(board, row, col):
    for i in range(row):
        if board[i] == col or abs(board[i] - col) == abs(i - row):
            return False
        
    return True

def solve_n_queens(solutions, board, row, n):
    if row == n:
        solutions.append(list(board))
        return

    for col in range(n):
        if is_safe(board, row, col):
            board[row] = col
            solve_n_queens(solutions, board, row + 1, n)
            board[row] = -1  # backtrack

def n_queens_backtracking(n):
    solutions = []
    board = [-1] * n
    solve_n_queens(solutions, board, 0, n)
    
    return solutions

def print_board(board):
    n = len(board)
    for i in range(n):
        for j in range(n):
            if board[i] == j:
                print("Q ", end="")
            else:
                print(". ", end="")
        print()
    print()


n = 4
solutions = n_queens_backtracking(n)
print("Number of solutions for", n, "-queens:", len(solutions))
for solution in solutions:
    print_board(solution)